docs/cheat-sheets/python-code-injection.mdx

```diff
@@ -0,0 +1,332 @@
+---
+slug: python-code-injection
+---
+
+import LinkToRegistryRule from "/src/components/LinkToRegistryRule"
+import CodeInjectionIntro from "/src/components/concept/_code-injection-intro.mdx"
+
+# Code injection prevention for Python
+
+<CodeInjectionIntro />
+
+## 1. Executing/evaluating code
+
+### 1.A. Executing code with **exec**
+
+`exec()` function supports dynamic execution of Python code.
+`exec()` can be dangerous if used to execute dynamic content.
+If this content can be input from outside the program, this may be a code injection vulnerability.
+
+Example:
+
+```python
+# Value supplied by user
+user_input = "');import requests;requests.get('localhost:3000');print('"
+
+# Vulnerable
+exec("foobar('{}')".format(user_input))
+```
+
+#### References
+
+- [**exec** documentation](https://docs.python.org/3/library/functions.html#exec)
+
+#### Mitigation
+
+Do not use `exec()` for non literal values. Alternatively:
+
+- Ensure executed content is not definable by external sources.
+- If it’s not possible, strip everything except alphanumeric characters from an input provided for the command string and arguments.
+- Don’t try to make `exec` safe with tricks like `{'__builtins__':{}}`
+
+#### Semgrep rule
+
+<LinkToRegistryRule ruleId="python.lang.security.audit.exec-detected.exec-detected" />
+
+### 1.B. Evaluating code with **eval**
+
+`eval()` function evaluates string value as a Python expression.
+`eval()` can be dangerous if used to evaluate dynamic content.
+If this content can be input from outside the program, this may be a code injection vulnerability.
```

Comment:
```suggestion
`eval()` function supports dynamic execution of Python code.
`eval()` can be dangerous if it is used to execute dynamic content (non literal content).
If this dynamic content has an input controllable by a user, it can cause a code injection vulnerability.
```